apps/src/codebridge/InfoPanel/ValidatedInstructions.tsx

```diff
@@ -0,0 +1,323 @@
+    } else if (hasNextLevel) {
+      // Props for a continue button.
+      return {
+        showNavigation,
+        navigationText: commonI18n.continue(),
+        handleNavigation: onNextPanel,
+        navigationIcon: {iconName: 'arrow-right', iconStyle: 'solid'},
+      };
+    } else {
+      // Props for a finish button.
+      return {
+        showNavigation,
+        navigationText: commonI18n.finish(),
+        handleNavigation: onFinish,
+      };
+    }
+  };
+
+  const renderTestButton = () => {
+    if (!hasConditions) {
+      return null;
+    }
+    return (
+      <div className={moduleStyles['bubble-' + theme]}>
+        {isTesting ? (
+          <Button
+            text={commonI18n.stopTests()}
+            onClick={handleStop}
+            color={'destructive'}
+            iconLeft={{iconStyle: 'solid', iconName: 'square'}}
+            className={moduleStyles.buttonInstruction}
+            size={'s'}
+          />
+        ) : (
+          <Button
+            text={commonI18n.test()}
+            onClick={() => handleTest()}
+            type={'secondary'}
+            disabled={isLoadingEnvironment}
+            iconLeft={{iconStyle: 'solid', iconName: 'flask'}}
+            className={moduleStyles.buttonInstruction}
+            color={'white'}
+            size={'s'}
+          />
+        )}
+      </div>
+    );
+  };
+
+  const {showNavigation, navigationText, navigationIcon, handleNavigation} =
+    getNavigationButtonProps();
+
+  // Don't render anything if we don't have any instructions.
+  if (instructionsText === undefined) {
+    return null;
+  }
+
+  return (
+    <div
+      id="instructions"
+      className={classNames(
+        moduleStyles['instructions-' + theme],
+        vertical && moduleStyles.vertical,
+        'instructions',
+        className
+      )}
+    >
+      <div
+        id="instructions-panel"
+        className={classNames(
+          moduleStyles.item,
+          vertical && moduleStyles.itemVertical
+        )}
+      >
+        {instructionsText && (
+          <div
+            key={instructionsText}
+            id="instructions-text"
+            className={classNames(
+              moduleStyles['bubble-' + theme],
+              moduleStyles.text
+            )}
+          >
+            {predictSettings?.isPredictLevel && <PredictSummary />}
+            <EnhancedSafeMarkdown
+              markdown={instructionsText}
+              className={moduleStyles.markdownText}
+              handleInstructionsTextClick={handleInstructionsTextClick}
+            />
+            <PredictQuestion
+              predictSettings={predictSettings}
+              predictResponse={predictResponse}
+              setPredictResponse={response =>
+                dispatch(setPredictResponse(response))
+              }
+              predictAnswerLocked={predictAnswerLocked}
+            />
+          </div>
+        )}
+        {SHOW_TEST_NAVIGATION_BUTTONS && renderTestButton()}
```

Comment:
could/should this be a component instead? Instead of a `renderTestButton()` function, it could be a component that accepts a `shouldShow` (or something) prop, ala:

```
const TestButton = ({ shouldDisplay = true }) => {
  if (!shouldDisplay) {
    return null
  }
  /* existing renderTestButton function as is */
}

// and then in the return here
<TextButton shouldDisplay = { SHOW_TEST_NAVIGATION_BUTTONS } />
```

They're functionally identical (though both the function version and the component version should probably be memoized in their respective ways), so it's just syntactic sugar.

Anyway, totally optional. I've just been trying to move myself away from these inline functions as much as possible so they tend to stick out to me.

Reply:
If we go with the button here long-term I'll refactor! For now I'm going to leave as-is since it's still in flux.